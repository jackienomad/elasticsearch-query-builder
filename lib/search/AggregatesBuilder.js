import { FieldMappingAdapter } from './FieldMappingAdapter'

export class AggregatesBuilder {
  static build(fields, fieldMapping) {
    const fieldAdapter =
      fieldMapping instanceof FieldMappingAdapter
        ? fieldMapping
        : new FieldMappingAdapter(fieldMapping)

    const aggregations = {}
    fields.forEach(field => {
      aggregations[field] = {
        terms: {
          field: fieldAdapter.getMappedField(field, true),
          size: 100
        }
      }
    })

    return aggregations
  }
}

export default AggregatesBuilder
