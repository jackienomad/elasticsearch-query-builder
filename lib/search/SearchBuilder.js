import {
  getMinimumByType,
  Operators,
  SearchTypes,
  MiscTypes
} from './searchUtils'
import { FieldMappingAdapter } from './FieldMappingAdapter'

export class SearchBuilder {
  /*------------------------------CONSTRUCTOR------------------------------*/
  constructor(index, type, fieldMapping) {
    this.index = index
    this.type = type
    this.fieldMapping = fieldMapping
    this.fieldMappingAdapter = new FieldMappingAdapter(fieldMapping)
    this.searchQuery = null
    this.mustFilterQuery = null
    this.notQuery = null
    this.sorting = null
  }

  destroy() {
    this.fieldMappingAdapter.destroy()

    this.index = null
    this.type = null
    this.fieldMapping = null
    this.fieldMappingAdapter = null
    this.searchQuery = null
    this.mustFilterQuery = null
    this.notQuery = null
    this.sorting = null
  }

  /*------------------------------METHODS------------------------------*/
  buildKeywords(keywords) {
    if (!this.notQuery) this.notQuery = []
    if (!this.searchQuery) this.searchQuery = []

    let queryGroup
    // build keywords
    keywords.forEach(({ field, operator, keyword, type }) => {
      const mappedField = this.fieldMappingAdapter.getMappedField(field)
      if (operator === Operators.NOT) {
        this.notQuery.push({
          match_phrase: { [mappedField]: keyword }
        })
      } else {
        let queryObject
        if (type === SearchTypes.EXACT) {
          queryObject = {
            match_phrase: { [mappedField]: keyword }
          }
        } else {
          queryObject = {
            match: {
              [mappedField]: {
                query: keyword,
                minimum_should_match: getMinimumByType(type)
              }
            }
          }
        }

        // init querygroup
        if (operator === Operators.AND || !queryGroup) {
          queryGroup = {
            bool: {
              should: []
            }
          }
          this.searchQuery.push(queryGroup)
        }

        queryGroup.bool.should.push(queryObject)
      }
    })
  }

  buildInclude(include) {
    const includeReturn = {}
    include.forEach(({ field, keyword }) => {
      const includedField = this.fieldMappingAdapter.getMappedField(field, true)
      if (!includeReturn[includedField]) includeReturn[includedField] = []
      includeReturn[includedField].push({
        match_phrase: { [includedField]: keyword }
      })
    })

    return Object.keys(includeReturn).map(field => ({
      bool: {
        should: includeReturn[field]
      }
    }))
  }

  buildExclude(exclude) {
    return exclude.map(({ field, keyword }) => ({
      match_phrase: {
        [this.fieldMappingAdapter.getMappedField(field)]: keyword
      }
    }))
  }

  buildMisc(misc) {
    if (!this.notQuery) this.notQuery = []
    if (!this.mustFilterQuery) this.mustFilterQuery = []
    if (!this.searchQuery) this.searchQuery = []

    misc &&
      misc.forEach(elem => {
        const miscField = this.fieldMappingAdapter.getMiscField(elem)

        if (
          elem === MiscTypes.BLANK_CONSIGNEES_EXCLUDED ||
          elem === MiscTypes.BLANK_SHIPPERS_EXCLUDED
        ) {
          this.searchQuery.push({
            exists: {
              field: miscField
            }
          })
        } else if (elem === MiscTypes.MASTER_SHIPMENTS_EXCLUDED) {
          this.notQuery.push({
            match_phrase: { [miscField]: 'Master' }
          })
        } else if (elem === MiscTypes.HOUSE_SHIPMENTS_ONLY) {
          this.mustFilterQuery.push({
            bool: {
              should: [
                {
                  match_phrase: { [miscField]: 'H' }
                },
                {
                  match_phrase: { [miscField]: 'House' }
                }
              ]
            }
          })
        } else if (elem === MiscTypes.MASTER_SHIPMENTS_ONLY) {
          this.mustFilterQuery.push({
            match_phrase: { [miscField]: 'Master' }
          })
        }
      })

    return {
      notQuery: this.notQuery,
      mustFilterQuery: this.mustFilterQuery,
      searchQuery: this.searchQuery
    }
  }

  buildSorting(sortBy) {
    let sortField = 'Arrival Date'
    let sortOrder = 'asc'
    this.sorting = []

    if (sortBy && sortBy.length > 0) {
      if (sortBy.charAt(0) === '-') {
        sortOrder = 'desc'
        sortField = this.fieldMappingAdapter.getMappedField(
          sortBy.substr(1),
          true
        )
      } else if (sortBy.charAt(0) === '+') {
        sortOrder = 'asc'
        sortField = this.fieldMappingAdapter.getMappedField(
          sortBy.substr(1),
          true
        )
      } else {
        sortOrder = 'asc'
        sortField = this.fieldMappingAdapter.getMappedField(sortBy, true)
      }
    }

    this.sorting.push(
      {
        [sortField]: {
          order: sortOrder
        }
      },
      '_score'
    )

    return this
  }

  buildQuery(
    keywords,
    include,
    exclude,
    misc,
    sortBy,
    limit,
    page,
    arrivalDates = ''
  ) {
    let [gte = '', lte = ''] = (arrivalDates && arrivalDates.split('-')) || []
    this.searchQuery = []
    this.mustFilterQuery = []
    this.notQuery = []

    this.buildKeywords(keywords)

    if (include) this.mustFilterQuery.push(...this.buildInclude(include))
    if (exclude) this.notQuery.push(...this.buildExclude(exclude))

    this.buildMisc(misc)
    this.buildSorting(sortBy)

    const querySearch = {
      size: limit,
      from: page * limit - limit,
      track_scores: true,
      sort: this.sorting,
      query: {
        bool: {
          must: this.searchQuery,
          must_not: this.notQuery,
          filter: [
            ...this.mustFilterQuery,
            {
              range: {
                'Arrival Date': {
                  gte: gte,
                  lte: lte
                }
              }
            }
          ]
        }
      }
    }

    return {
      index: this.index,
      type: this.type,
      body: querySearch
    }
  }
}

export default SearchBuilder
