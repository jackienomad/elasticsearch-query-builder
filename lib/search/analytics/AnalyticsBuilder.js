const buildDisplayCount = (analyticsAggregateFields, fieldMapper) => {
  const aggregateDisplay = {}

  Object.keys(analyticsAggregateFields).forEach(fieldName => {
    if (fieldName !== 'shipments_count') {
      const { mappedField } = analyticsAggregateFields[fieldName]

      const { field } = fieldMapper[mappedField]
      aggregateDisplay[fieldName] = {
        sum: {
          field
        }
      }
    }
  })

  return aggregateDisplay
}

const getAnalyzedField = (field, fieldMapper, analyzeField) => {
  const getField = analyzeField[field].mappedField
  return fieldMapper[getField].not_analyzed_field
}

const buildSorting = (sortBy, analyticsAggregateFields) => {
  let sortOrder = 'desc'
  let sortName = ''
  if (sortBy && sortBy.length > 0) {
    switch (sortBy.charAt(0)) {
      case '-':
        sortOrder = 'desc'
        sortName = sortBy.substr(1).toLowerCase()
        break
      case '+':
        sortOrder = 'asc'
        sortName = sortBy.substr(1).toLowerCase()

        break
      default:
        sortOrder = 'desc'
        sortName = sortBy.toLowerCase()
    }

    if (sortName === 'shipments_count') {
      sortName = analyticsAggregateFields[sortName].mappedField
    }

    return [{ [sortName]: `${sortOrder}` }]
  }

  return null
}

export class AnalyticsBuilder {
  constructor(searchBuilder, searchAnalyticsFieldMapping) {
    if (!searchBuilder) throw new Error('searchBuilder is required')
    if (typeof searchAnalyticsFieldMapping !== 'object') {
      throw new Error('searchAnalyticsFieldMapping is required')
    }

    this.searchBuilder = searchBuilder
    this.searchAggregatedFieldMapping = searchAnalyticsFieldMapping

    const {
      analyticsAggregateFields,
      analyticsAnalyzeFields,
      analyticsDefaultDisplaySortBy
    } = searchAnalyticsFieldMapping

    this.analyticsAggregateFields = analyticsAggregateFields
    this.analyzeFields = analyticsAnalyzeFields
    this.defaultSortBy = analyticsDefaultDisplaySortBy
  }

  destroy() {
    this.searchBuilder = null
    this.searchAggregatedFieldMapping = null
    this.analyticsAggregateFields = null
    this.analyzeFields = null
    this.defaultSortBy = null
  }

  buildQuery(
    analyzedField,
    keywords,
    include,
    exclude,
    misc,
    sortBy,
    size,
    arrivalDate = ''
  ) {
    const query = this.searchBuilder.buildQuery(
      keywords,
      include,
      exclude,
      misc,
      [],
      0,
      0,
      arrivalDate
    )

    const aggregateDisplay = buildDisplayCount(
      this.analyticsAggregateFields,
      this.searchAggregatedFieldMapping
    )

    if (sortBy === '') {
      sortBy = this.defaultSortBy
    }

    const sorting = buildSorting(sortBy, this.analyticsAggregateFields)

    const aggregation = {
      shipmentAggregate: {
        terms: {
          field: getAnalyzedField(
            analyzedField,
            this.searchAggregatedFieldMapping,
            this.analyzeFields
          ),
          size: size
        },
        aggs: { ...aggregateDisplay }
      }
    }

    if (sorting != null) {
      aggregation.shipmentAggregate.terms['order'] = sorting
    }

    query.body.size = 0
    query.body.from = 0
    delete query.body.sort

    query.body['aggs'] = aggregation
    return query
  }
}

export default AnalyticsBuilder
