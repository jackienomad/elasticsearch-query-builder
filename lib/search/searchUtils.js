export const refinementIndexes = ['keyword', 'field']
export const keywordIndexes = ['keyword', 'operator', 'type', 'field']

export const Operators = {
  AND: 'and',
  OR: 'or',
  NOT: 'not'
}
export const SearchTypes = {
  EXACT: 'exact',
  FUZZY: 'fuzzy',
  REGULAR: 'regular'
}
export const MiscTypes = {
  BLANK_CONSIGNEES_EXCLUDED: 'BlankConsigneesExcluded',
  BLANK_SHIPPERS_EXCLUDED: 'BlankShippersExcluded',
  MASTER_SHIPMENTS_EXCLUDED: 'MasterShipmentsExcluded',
  HOUSE_SHIPMENTS_ONLY: 'HouseShipmentsOnly',
  MASTER_SHIPMENTS_ONLY: 'MasterShipmentsOnly'
}

export const percentByType = {
  [SearchTypes.REGULAR]: '100%',
  [SearchTypes.FUZZY]: '75%'
}

export function getMinimumByType(type) {
  return percentByType[type]
}

export function transformArray(transformee, indexes) {
  if (transformee === undefined) {
    return []
  }
  return transformee.map(val => {
    let values = val.split(',')

    if (values.length > indexes.length) {
      const excess = values.splice(
        values.length - indexes.length + 1,
        values.length
      )
      values = [values.join(','), ...excess]
    }

    const obj = {}
    indexes.forEach((prop, index) => {
      obj[prop] = values[index]
    })
    return obj
  })
}
