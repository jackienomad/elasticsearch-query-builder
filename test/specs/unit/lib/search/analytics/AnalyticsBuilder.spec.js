import fieldMapping from '../dataset/sampleFieldMapping'
import { AnalyticsBuilder } from 'qb/@/search/analytics/AnalyticsBuilder'
import {
  analyticsAggregateFields,
  analyticsAnalyzeFields,
  analyticsDefaultDisplaySortBy,
  analyticsDefaultValueShow
} from '../dataset/analyticsAggregateFields'
import { SearchBuilder } from 'qb/@/search/SearchBuilder'
import Faker from 'faker'
import { MiscTypes } from 'qb/@/search/searchUtils'

import * as expectedReturnQueryBuilder from './expectedReturnQueryBuilder'

describe('Search Analytics AnalyticsBuilder', () => {
  describe('constructor and destructor', () => {
    it('should be instantiable', () => {
      expect(() => {
        const searchBuilder = new SearchBuilder(
          Faker.random.word(),
          Faker.random.word(),
          fieldMapping
        )
        const merged = Object.assign(fieldMapping, analyticsAggregateFields)

        new AnalyticsBuilder(searchBuilder, merged)
      }).not.toThrowError()
    })

    it('should be destroyable', () => {
      const searchBuilder = new SearchBuilder(
        Faker.random.word(),
        Faker.random.word(),
        fieldMapping
      )

      const merged = Object.assign(fieldMapping, analyticsAggregateFields)
      const builder = new AnalyticsBuilder(searchBuilder, merged)
      expect(() => {
        builder.destroy()
      }).not.toThrowError()
    })
  })

  describe('build a query and aggregate term ', () => {
    it('should build a query for the analytics as a aggregate term consignees', () => {
      const searchBuilder = new SearchBuilder(
        'us_import',
        'shipments',
        fieldMapping
      )
      const merged = Object.assign(
        fieldMapping,
        analyticsAggregateFields,
        analyticsAnalyzeFields,
        analyticsDefaultDisplaySortBy,
        analyticsDefaultValueShow
      )

      const builder = new AnalyticsBuilder(searchBuilder, merged)

      const build = builder.buildQuery(
        'consignees',
        [
          {
            field: 'Consignee',
            keyword: 'apple',
            operator: 'and',
            type: 'fuzzy'
          }
        ],
        [],
        [],
        [MiscTypes.BLANK_CONSIGNEES_EXCLUDED],
        '',
        5,
        '1512938867-1544475436'
      )

      expect(build).toMatchSnapshot()
    })

    it('should build a query for the analytics as a aggregate term consignees', () => {
      const searchBuilder = new SearchBuilder(
        'us_import',
        'shipments',
        fieldMapping
      )
      const merged = Object.assign(
        fieldMapping,
        analyticsAggregateFields,
        analyticsAnalyzeFields
      )

      const builder = new AnalyticsBuilder(searchBuilder, merged)

      const build = builder.buildQuery(
        'consignees',
        [
          {
            field: 'Consignee',
            keyword: 'apple',
            operator: 'and',
            type: 'fuzzy'
          }
        ],
        [],
        [],
        [MiscTypes.BLANK_CONSIGNEES_EXCLUDED],
        '',
        5,
        '1512938867-1544475436'
      )

      expect(build).toMatchSnapshot()
    })

    it('should build a query for the analytics as a aggregate term states', () => {
      const searchBuilder = new SearchBuilder(
        'us_import',
        'shipments',
        fieldMapping
      )
      const merged = Object.assign(
        fieldMapping,
        analyticsAggregateFields,
        analyticsAnalyzeFields
      )

      const builder = new AnalyticsBuilder(searchBuilder, merged)

      const build = builder.buildQuery(
        'states',
        [
          {
            field: 'Consignee',
            keyword: 'apple',
            operator: 'and',
            type: 'fuzzy'
          }
        ],
        [],
        [],
        [MiscTypes.BLANK_CONSIGNEES_EXCLUDED],
        '',
        5,
        '1512938867-1544475436'
      )

      expect(build).toMatchSnapshot()
    })

    it('should build a query for the analytics as a aggregate term shippers', () => {
      const searchBuilder = new SearchBuilder(
        'us_import',
        'shipments',
        fieldMapping
      )
      const merged = Object.assign(
        fieldMapping,
        analyticsAggregateFields,
        analyticsAnalyzeFields
      )

      const builder = new AnalyticsBuilder(searchBuilder, merged)

      const build = builder.buildQuery(
        'shippers',
        [
          {
            field: 'Consignee',
            keyword: 'apple',
            operator: 'and',
            type: 'fuzzy'
          }
        ],
        [],
        [],
        [MiscTypes.BLANK_CONSIGNEES_EXCLUDED],
        '',
        5,
        '1512938867-1544475436'
      )

      expect(build).toMatchSnapshot()
    })

    it('should build a query for the analytics as a aggregate term us_ports', () => {
      const searchBuilder = new SearchBuilder(
        'us_import',
        'shipments',
        fieldMapping
      )
      const merged = Object.assign(
        fieldMapping,
        analyticsAggregateFields,
        analyticsAnalyzeFields
      )

      const builder = new AnalyticsBuilder(searchBuilder, merged)

      const build = builder.buildQuery(
        'us_ports',
        [
          {
            field: 'Consignee',
            keyword: 'apple',
            operator: 'and',
            type: 'fuzzy'
          }
        ],
        [],
        [],
        [MiscTypes.BLANK_CONSIGNEES_EXCLUDED],
        '',
        5,
        '1512938867-1544475436'
      )

      expect(build).toMatchSnapshot()
    })

    it('should build a query for the analytics as a aggregate term foreign_ports', () => {
      const searchBuilder = new SearchBuilder(
        'us_import',
        'shipments',
        fieldMapping
      )
      const merged = Object.assign(
        fieldMapping,
        analyticsAggregateFields,
        analyticsAnalyzeFields
      )

      const builder = new AnalyticsBuilder(searchBuilder, merged)

      const build = builder.buildQuery(
        'foreign_ports',
        [
          {
            field: 'Consignee',
            keyword: 'apple',
            operator: 'and',
            type: 'fuzzy'
          }
        ],
        [],
        [],
        [MiscTypes.BLANK_CONSIGNEES_EXCLUDED],
        '',
        5,
        '1512938867-1544475436'
      )

      expect(build).toMatchSnapshot()
    })

    it('should build a query for the analytics as a aggregate term origin_countries', () => {
      const searchBuilder = new SearchBuilder(
        'us_import',
        'shipments',
        fieldMapping
      )
      const merged = Object.assign(
        fieldMapping,
        analyticsAggregateFields,
        analyticsAnalyzeFields
      )

      const builder = new AnalyticsBuilder(searchBuilder, merged)

      const build = builder.buildQuery(
        'origin_countries',
        [
          {
            field: 'Consignee',
            keyword: 'apple',
            operator: 'and',
            type: 'fuzzy'
          }
        ],
        [],
        [],
        [MiscTypes.BLANK_CONSIGNEES_EXCLUDED],
        '',
        5,
        '1512938867-1544475436'
      )

      expect(build).toMatchSnapshot()
    })

    it('should sort by shipments_count desc (analyticsAggregateFields)', () => {
      const searchBuilder = new SearchBuilder(
        'us_import',
        'shipments',
        fieldMapping
      )
      const merged = Object.assign(
        fieldMapping,
        analyticsAggregateFields,
        analyticsAnalyzeFields
      )
      const builder = new AnalyticsBuilder(searchBuilder, merged)
      const build = builder.buildQuery(
        'consignees',
        [
          {
            field: 'Consignee',
            keyword: 'apple',
            operator: 'and',
            type: 'fuzzy'
          }
        ],
        [],
        [],
        [MiscTypes.BLANK_CONSIGNEES_EXCLUDED],
        `-shipments_count`,
        5,
        '1512938867-1544475436'
      )

      expect(build).toMatchSnapshot()
    })

    it('should work without the sortBy and get the default sortBy', () => {
      const searchBuilder = new SearchBuilder(
        'us_import',
        'shipments',
        fieldMapping
      )
      const merged = Object.assign(
        fieldMapping,
        analyticsAggregateFields,
        analyticsAnalyzeFields,
        analyticsDefaultDisplaySortBy
      )
      const builder = new AnalyticsBuilder(searchBuilder, merged)

      const build = builder.buildQuery(
        'consignees',
        [
          {
            field: 'Consignee',
            keyword: 'apple',
            operator: 'and',
            type: 'fuzzy'
          }
        ],
        [],
        [],
        [MiscTypes.BLANK_CONSIGNEES_EXCLUDED],
        '',
        5,
        '1512938867-1544475436'
      )
      const { body } = build
      const { aggs } = body
      const { shipmentAggregate } = aggs
      const { terms } = shipmentAggregate
      const { order } = terms

      expect(order).toEqual(
        expectedReturnQueryBuilder.responseOfQueryBuilderTermsOrderDefault
      )
    })

    it('should sort by according to the default field (analyticsAggregateFields)', () => {
      const searchBuilder = new SearchBuilder(
        'us_import',
        'shipments',
        fieldMapping
      )
      const merged = Object.assign(
        fieldMapping,
        analyticsAggregateFields,
        analyticsAnalyzeFields
      )
      const builder = new AnalyticsBuilder(searchBuilder, merged)

      const build = builder.buildQuery(
        'consignees',
        [
          {
            field: 'Consignee',
            keyword: 'apple',
            operator: 'and',
            type: 'fuzzy'
          }
        ],
        [],
        [],
        [MiscTypes.BLANK_CONSIGNEES_EXCLUDED],
        `weight_lb`,
        5,
        '1512938867-1544475436'
      )

      expect(build).toMatchSnapshot()
    })

    it('should throw an error on missing fieldMapping', () => {
      expect(() => {
        new AnalyticsBuilder()
      }).toThrowError('searchBuilder is required')

      expect(() => {
        const searchBuilder = new SearchBuilder(
          'us_import',
          'shipments',
          fieldMapping
        )
        new AnalyticsBuilder(searchBuilder)
      }).toThrowError('searchAnalyticsFieldMapping is required')
    })
  })
})
