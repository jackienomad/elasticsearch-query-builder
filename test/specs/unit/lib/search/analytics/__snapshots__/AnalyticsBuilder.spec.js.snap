// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Analytics AnalyticsBuilder build a query and aggregate term  should build a query for the analytics as a aggregate term consignees 1`] = `
Object {
  "body": Object {
    "aggs": Object {
      "shipmentAggregate": Object {
        "aggs": Object {
          "containers_count": Object {
            "sum": Object {
              "field": "Number of Containers",
            },
          },
          "weight_kg": Object {
            "sum": Object {
              "field": "Weight",
            },
          },
          "weight_lb": Object {
            "sum": Object {
              "field": "Weight",
            },
          },
        },
        "terms": Object {
          "field": "Consignee.Name.untouched",
          "order": Array [
            Object {
              "_count": "desc",
            },
          ],
          "size": 5,
        },
      },
    },
    "from": 0,
    "query": Object {
      "bool": Object {
        "filter": Array [
          Object {
            "range": Object {
              "Arrival Date": Object {
                "gte": "1512938867",
                "lte": "1544475436",
              },
            },
          },
        ],
        "must": Array [
          Object {
            "bool": Object {
              "should": Array [
                Object {
                  "match": Object {
                    "Consignee.Name": Object {
                      "minimum_should_match": "75%",
                      "query": "apple",
                    },
                  },
                },
              ],
            },
          },
          Object {
            "exists": Object {
              "field": "Consignee.Name",
            },
          },
        ],
        "must_not": Array [],
      },
    },
    "size": 0,
    "track_scores": true,
  },
  "index": "us_import",
  "type": "shipments",
}
`;

exports[`Search Analytics AnalyticsBuilder build a query and aggregate term  should build a query for the analytics as a aggregate term consignees 2`] = `
Object {
  "body": Object {
    "aggs": Object {
      "shipmentAggregate": Object {
        "aggs": Object {
          "containers_count": Object {
            "sum": Object {
              "field": "Number of Containers",
            },
          },
          "weight_kg": Object {
            "sum": Object {
              "field": "Weight",
            },
          },
          "weight_lb": Object {
            "sum": Object {
              "field": "Weight",
            },
          },
        },
        "terms": Object {
          "field": "Consignee.Name.untouched",
          "order": Array [
            Object {
              "_count": "desc",
            },
          ],
          "size": 5,
        },
      },
    },
    "from": 0,
    "query": Object {
      "bool": Object {
        "filter": Array [
          Object {
            "range": Object {
              "Arrival Date": Object {
                "gte": "1512938867",
                "lte": "1544475436",
              },
            },
          },
        ],
        "must": Array [
          Object {
            "bool": Object {
              "should": Array [
                Object {
                  "match": Object {
                    "Consignee.Name": Object {
                      "minimum_should_match": "75%",
                      "query": "apple",
                    },
                  },
                },
              ],
            },
          },
          Object {
            "exists": Object {
              "field": "Consignee.Name",
            },
          },
        ],
        "must_not": Array [],
      },
    },
    "size": 0,
    "track_scores": true,
  },
  "index": "us_import",
  "type": "shipments",
}
`;

exports[`Search Analytics AnalyticsBuilder build a query and aggregate term  should build a query for the analytics as a aggregate term foreign_ports 1`] = `
Object {
  "body": Object {
    "aggs": Object {
      "shipmentAggregate": Object {
        "aggs": Object {
          "containers_count": Object {
            "sum": Object {
              "field": "Number of Containers",
            },
          },
          "weight_kg": Object {
            "sum": Object {
              "field": "Weight",
            },
          },
          "weight_lb": Object {
            "sum": Object {
              "field": "Weight",
            },
          },
        },
        "terms": Object {
          "field": "Foreign Port of Lading.untouched",
          "order": Array [
            Object {
              "_count": "desc",
            },
          ],
          "size": 5,
        },
      },
    },
    "from": 0,
    "query": Object {
      "bool": Object {
        "filter": Array [
          Object {
            "range": Object {
              "Arrival Date": Object {
                "gte": "1512938867",
                "lte": "1544475436",
              },
            },
          },
        ],
        "must": Array [
          Object {
            "bool": Object {
              "should": Array [
                Object {
                  "match": Object {
                    "Consignee.Name": Object {
                      "minimum_should_match": "75%",
                      "query": "apple",
                    },
                  },
                },
              ],
            },
          },
          Object {
            "exists": Object {
              "field": "Consignee.Name",
            },
          },
        ],
        "must_not": Array [],
      },
    },
    "size": 0,
    "track_scores": true,
  },
  "index": "us_import",
  "type": "shipments",
}
`;

exports[`Search Analytics AnalyticsBuilder build a query and aggregate term  should build a query for the analytics as a aggregate term origin_countries 1`] = `
Object {
  "body": Object {
    "aggs": Object {
      "shipmentAggregate": Object {
        "aggs": Object {
          "containers_count": Object {
            "sum": Object {
              "field": "Number of Containers",
            },
          },
          "weight_kg": Object {
            "sum": Object {
              "field": "Weight",
            },
          },
          "weight_lb": Object {
            "sum": Object {
              "field": "Weight",
            },
          },
        },
        "terms": Object {
          "field": "Country of Origin.untouched",
          "order": Array [
            Object {
              "_count": "desc",
            },
          ],
          "size": 5,
        },
      },
    },
    "from": 0,
    "query": Object {
      "bool": Object {
        "filter": Array [
          Object {
            "range": Object {
              "Arrival Date": Object {
                "gte": "1512938867",
                "lte": "1544475436",
              },
            },
          },
        ],
        "must": Array [
          Object {
            "bool": Object {
              "should": Array [
                Object {
                  "match": Object {
                    "Consignee.Name": Object {
                      "minimum_should_match": "75%",
                      "query": "apple",
                    },
                  },
                },
              ],
            },
          },
          Object {
            "exists": Object {
              "field": "Consignee.Name",
            },
          },
        ],
        "must_not": Array [],
      },
    },
    "size": 0,
    "track_scores": true,
  },
  "index": "us_import",
  "type": "shipments",
}
`;

exports[`Search Analytics AnalyticsBuilder build a query and aggregate term  should build a query for the analytics as a aggregate term shippers 1`] = `
Object {
  "body": Object {
    "aggs": Object {
      "shipmentAggregate": Object {
        "aggs": Object {
          "containers_count": Object {
            "sum": Object {
              "field": "Number of Containers",
            },
          },
          "weight_kg": Object {
            "sum": Object {
              "field": "Weight",
            },
          },
          "weight_lb": Object {
            "sum": Object {
              "field": "Weight",
            },
          },
        },
        "terms": Object {
          "field": "Shipper.Name.untouched",
          "order": Array [
            Object {
              "_count": "desc",
            },
          ],
          "size": 5,
        },
      },
    },
    "from": 0,
    "query": Object {
      "bool": Object {
        "filter": Array [
          Object {
            "range": Object {
              "Arrival Date": Object {
                "gte": "1512938867",
                "lte": "1544475436",
              },
            },
          },
        ],
        "must": Array [
          Object {
            "bool": Object {
              "should": Array [
                Object {
                  "match": Object {
                    "Consignee.Name": Object {
                      "minimum_should_match": "75%",
                      "query": "apple",
                    },
                  },
                },
              ],
            },
          },
          Object {
            "exists": Object {
              "field": "Consignee.Name",
            },
          },
        ],
        "must_not": Array [],
      },
    },
    "size": 0,
    "track_scores": true,
  },
  "index": "us_import",
  "type": "shipments",
}
`;

exports[`Search Analytics AnalyticsBuilder build a query and aggregate term  should build a query for the analytics as a aggregate term states 1`] = `
Object {
  "body": Object {
    "aggs": Object {
      "shipmentAggregate": Object {
        "aggs": Object {
          "containers_count": Object {
            "sum": Object {
              "field": "Number of Containers",
            },
          },
          "weight_kg": Object {
            "sum": Object {
              "field": "Weight",
            },
          },
          "weight_lb": Object {
            "sum": Object {
              "field": "Weight",
            },
          },
        },
        "terms": Object {
          "field": "Consignee.Cleaned Address Breakdown.StateName.untouched",
          "order": Array [
            Object {
              "_count": "desc",
            },
          ],
          "size": 5,
        },
      },
    },
    "from": 0,
    "query": Object {
      "bool": Object {
        "filter": Array [
          Object {
            "range": Object {
              "Arrival Date": Object {
                "gte": "1512938867",
                "lte": "1544475436",
              },
            },
          },
        ],
        "must": Array [
          Object {
            "bool": Object {
              "should": Array [
                Object {
                  "match": Object {
                    "Consignee.Name": Object {
                      "minimum_should_match": "75%",
                      "query": "apple",
                    },
                  },
                },
              ],
            },
          },
          Object {
            "exists": Object {
              "field": "Consignee.Name",
            },
          },
        ],
        "must_not": Array [],
      },
    },
    "size": 0,
    "track_scores": true,
  },
  "index": "us_import",
  "type": "shipments",
}
`;

exports[`Search Analytics AnalyticsBuilder build a query and aggregate term  should build a query for the analytics as a aggregate term us_ports 1`] = `
Object {
  "body": Object {
    "aggs": Object {
      "shipmentAggregate": Object {
        "aggs": Object {
          "containers_count": Object {
            "sum": Object {
              "field": "Number of Containers",
            },
          },
          "weight_kg": Object {
            "sum": Object {
              "field": "Weight",
            },
          },
          "weight_lb": Object {
            "sum": Object {
              "field": "Weight",
            },
          },
        },
        "terms": Object {
          "field": "Port of Unlading.untouched",
          "order": Array [
            Object {
              "_count": "desc",
            },
          ],
          "size": 5,
        },
      },
    },
    "from": 0,
    "query": Object {
      "bool": Object {
        "filter": Array [
          Object {
            "range": Object {
              "Arrival Date": Object {
                "gte": "1512938867",
                "lte": "1544475436",
              },
            },
          },
        ],
        "must": Array [
          Object {
            "bool": Object {
              "should": Array [
                Object {
                  "match": Object {
                    "Consignee.Name": Object {
                      "minimum_should_match": "75%",
                      "query": "apple",
                    },
                  },
                },
              ],
            },
          },
          Object {
            "exists": Object {
              "field": "Consignee.Name",
            },
          },
        ],
        "must_not": Array [],
      },
    },
    "size": 0,
    "track_scores": true,
  },
  "index": "us_import",
  "type": "shipments",
}
`;

exports[`Search Analytics AnalyticsBuilder build a query and aggregate term  should sort by according to the default field (analyticsAggregateFields) 1`] = `
Object {
  "body": Object {
    "aggs": Object {
      "shipmentAggregate": Object {
        "aggs": Object {
          "containers_count": Object {
            "sum": Object {
              "field": "Number of Containers",
            },
          },
          "weight_kg": Object {
            "sum": Object {
              "field": "Weight",
            },
          },
          "weight_lb": Object {
            "sum": Object {
              "field": "Weight",
            },
          },
        },
        "terms": Object {
          "field": "Consignee.Name.untouched",
          "order": Array [
            Object {
              "weight_lb": "desc",
            },
          ],
          "size": 5,
        },
      },
    },
    "from": 0,
    "query": Object {
      "bool": Object {
        "filter": Array [
          Object {
            "range": Object {
              "Arrival Date": Object {
                "gte": "1512938867",
                "lte": "1544475436",
              },
            },
          },
        ],
        "must": Array [
          Object {
            "bool": Object {
              "should": Array [
                Object {
                  "match": Object {
                    "Consignee.Name": Object {
                      "minimum_should_match": "75%",
                      "query": "apple",
                    },
                  },
                },
              ],
            },
          },
          Object {
            "exists": Object {
              "field": "Consignee.Name",
            },
          },
        ],
        "must_not": Array [],
      },
    },
    "size": 0,
    "track_scores": true,
  },
  "index": "us_import",
  "type": "shipments",
}
`;

exports[`Search Analytics AnalyticsBuilder build a query and aggregate term  should sort by shipments_count desc (analyticsAggregateFields) 1`] = `
Object {
  "body": Object {
    "aggs": Object {
      "shipmentAggregate": Object {
        "aggs": Object {
          "containers_count": Object {
            "sum": Object {
              "field": "Number of Containers",
            },
          },
          "weight_kg": Object {
            "sum": Object {
              "field": "Weight",
            },
          },
          "weight_lb": Object {
            "sum": Object {
              "field": "Weight",
            },
          },
        },
        "terms": Object {
          "field": "Consignee.Name.untouched",
          "order": Array [
            Object {
              "_count": "desc",
            },
          ],
          "size": 5,
        },
      },
    },
    "from": 0,
    "query": Object {
      "bool": Object {
        "filter": Array [
          Object {
            "range": Object {
              "Arrival Date": Object {
                "gte": "1512938867",
                "lte": "1544475436",
              },
            },
          },
        ],
        "must": Array [
          Object {
            "bool": Object {
              "should": Array [
                Object {
                  "match": Object {
                    "Consignee.Name": Object {
                      "minimum_should_match": "75%",
                      "query": "apple",
                    },
                  },
                },
              ],
            },
          },
          Object {
            "exists": Object {
              "field": "Consignee.Name",
            },
          },
        ],
        "must_not": Array [],
      },
    },
    "size": 0,
    "track_scores": true,
  },
  "index": "us_import",
  "type": "shipments",
}
`;
